src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -54,17 +54,35 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
     workInProgress.pendingWorkPriority = NoWork;
   }
 
-  function updateClassComponent(current, workInProgress) {
+  function updateClassComponent(current : ?Fiber, workInProgress : Fiber) {
     var props = workInProgress.pendingProps;
-    if (!workInProgress.stateNode) {
+    var instance = workInProgress.stateNode;
+    if (!instance) {
       var ctor = workInProgress.type;
-      workInProgress.stateNode = new ctor(props);
+      workInProgress.stateNode = instance = new ctor(props);
+    } else if (typeof instance.shouldComponentUpdate === 'function') {
+      if (current && current.memoizedProps) {
+        // Revert to the last flushed props, incase we aborted an update.
+        instance.props = current.memoizedProps;
+        if (!instance.shouldComponentUpdate(props)) {
+          return bailoutOnCurrent(current, workInProgress);
+        }
+      }
+      if (!workInProgress.hasWorkInProgress && workInProgress.memoizedProps) {
```

Comment:
What do you mean by exhaustive? They're not covering all cases since they don't cover the initial mount case.

The above case is basically what we already have in the stack based React. If the props are on the thing that we last rendered onto the screen we can just bail out.

The second statement is a new thing.

The work in progress can be in one of three states. a) Its children were previously in the progress of reconciling, but were interrupted. If that's the case, we can't safely reuse the work because this subtree is in an inconsistent state. b) This work was already started earlier AND completed all its children but then something *else* was aborted. We eventually came back here but we can reuse the work. c) This workInProgress is not actually a new clone, but a reused object. That object happens to have the state of the props in the state *before* the current one on it. So we might as well reuse that for free. This is the ping-pong case. If you alternate between two different props.

The second statement covers b and c.